chain/network/src/peer/stream.rs

```diff
@@ -0,0 +1,203 @@
+
+type ReadHalf = tokio::io::ReadHalf<tokio::net::TcpStream>;
+type WriteHalf = tokio::io::WriteHalf<tokio::net::TcpStream>;
+
+#[derive(thiserror::Error, Debug)]
+pub(crate) enum SendError {
+    #[error("IO error")]
+    IO(#[source] io::Error),
+    #[error("queue is full, got {got_bytes}B, max capacity is {want_max_bytes}")]
+    QueueOverflow { got_bytes: usize, want_max_bytes: usize },
+}
+
+#[derive(thiserror::Error, Debug)]
+pub(crate) enum RecvError {
+    #[error("IO error")]
+    IO(#[source] io::Error),
+    #[error("message too large: got {got_bytes}B, want <={want_max_bytes}B")]
+    MessageTooLarge { got_bytes: usize, want_max_bytes: usize },
+}
+
+#[derive(actix::Message)]
+#[rtype(result = "()")]
+pub(crate) struct Frame(pub Vec<u8>);
+
+/// Event loop receiving and processing messages.
+/// Loop waits for the message to be processed before reading the next message.
+/// Note that if the message handler spawns an asynchronous subhandler and returns,
+/// then the loop will start reading the next message before the subhandler returns.
+/// Loop uses a fixed small buffer allocated by BufReader.
+/// For each message it allocates a Vec with exact size of the message.
+// TODO(gprusak): once borsh support is dropped, we can parse a proto
+// directly from the stream.
+async fn run_recv_loop(
+    peer_addr: SocketAddr,
+    read: ReadHalf,
+    addr: actix::Addr<PeerActor>,
+    stats: Arc<connection::Stats>,
+) -> Result<(), RecvError> {
+    const READ_BUFFER_CAPACITY: usize = 8 * 1024;
+    let mut read = tokio::io::BufReader::with_capacity(READ_BUFFER_CAPACITY, read);
+
+    let msg_size_metric =
+        metrics::MetricGuard::new(&metrics::PEER_MSG_SIZE_BYTES, vec![peer_addr.to_string()]);
+    let buf_size_metric = metrics::MetricGuard::new(
+        &metrics::PEER_DATA_READ_BUFFER_SIZE,
+        vec![peer_addr.to_string()],
+    );
+    loop {
+        let n = read.read_u32_le().await.map_err(RecvError::IO)? as usize;
+        if n > NETWORK_MESSAGE_MAX_SIZE_BYTES {
+            return Err(RecvError::MessageTooLarge {
+                got_bytes: n,
+                want_max_bytes: NETWORK_MESSAGE_MAX_SIZE_BYTES,
+            });
+        }
+        msg_size_metric.observe(n as f64);
+        buf_size_metric.set(n as i64);
+        let mut buf = vec![0; n];
+        let t = metrics::PEER_MSG_READ_LATENCY.start_timer();
+        read.read_exact(&mut buf[..]).await.map_err(RecvError::IO)?;
+        t.observe_duration();
+        buf_size_metric.set(0);
+        stats.received_messages.fetch_add(1, Ordering::Relaxed);
+        stats.received_bytes.fetch_add(n as u64, Ordering::Relaxed);
+        if let Err(_) = addr.send(Frame(buf)).await {
+            // We got mailbox error, which means that Actor has stopped,
+            // so we should just close the stream.
+            return Ok(());
+        }
+    }
+}
+async fn run_send_loop(
+    tcp_send: WriteHalf,
+    mut queue_recv: tokio::sync::mpsc::UnboundedReceiver<Vec<u8>>,
```

Comment:
Feels a bit unsymmetric that we don't wrap Vec into a Frame here, but not that it matters.

Reply:
done